internal/reflect/interfaces.go

```diff
@@ -74,18 +75,38 @@ func FromUnknownable(ctx context.Context, typ attr.Type, val Unknownable, path p
 	if val.GetUnknown(ctx) {
 		tfVal := tftypes.NewValue(typ.TerraformType(ctx), tftypes.UnknownValue)
 
-		if typeWithValidate, ok := typ.(xattr.TypeWithValidate); ok {
-			diags.Append(typeWithValidate.Validate(ctx, tfVal, path)...)
+		res, err := typ.ValueFromTerraform(ctx, tfVal)
+		if err != nil {
+			return nil, append(diags, valueFromTerraformErrorDiag(err, path))
+		}
+
+		switch t := res.(type) {
+		case xattr.ValidateableAttribute:
```

Comment:
> For example, in the case of a custom type only being used on a function parameter, it might not implement this ValidateableAttribute interface. When the original ArgumentData is being built from tfplugin-go, it would call the ValidateableParameter interface, but when calling (ArgumentData).Get or (ResultData).Set it would be using the ValidateableAttribute interface (via the reflection) and therefore not get called.

If a custom type that is only being used as a function parameter implements `ValidateableParameter` but does not implement `ValidateableAttribute`, the call to `ValidateParameter()` will occur during the `CallFunction` RPC when `fromproto5/6.ArgumentsData()` is called. The call to `fromproto5/6.ArgumentsData()` occurs, prior to calling `(ArgumentData).Get()` or `(ResultData).Set()`, the latter of which are typically called within the `Run()` method of the provider-defined function. Does this answer your concern?

> I wonder how big of a refactor it would be to completely remove validation from the reflect package and make it the responsibility of the caller to validate the reflection value is valid. So (ArgumentData).Get, (ResultData).Set, (Data).Get, (Data).Set, etc. would need to perform this validation.

I chatted with Brian about this out-of-band, as I was also wondering about the embedded validation within reflection. I think this is worthy of further investigation but I believe would be best suited as a follow-up issue/discussion.

> Or perhaps we should be attempting to add the parameter validation into the reflect logic, although we'll likely run into import cycles there because of FuncError, so maybe if we don't do anything here it's just a documentation problem?

In terms of the documentation, is the concern that it may not be clear under which circumstances to implement `ValidateableParameter` vs `ValidateableAttribute`? There's some updates to the docs in this PR to try and make this clear, but let me know if you think this should be expanded.

Reply:
This is an interesting case! If we want to explicitly validate values that are being passed to `Result.Set` it seems that maybe neither `ValidateableParameter` nor `ValidateableAttribute` are appropriate interfaces, as the value is neither a parameter nor an attribute.

One option might be to consider a `Validateable` interface that operates solely on the value. Perhaps something along the following lines:

```go
type Validateable interface {
	Validate(context.Context, ValidateRequest, *ValidateResponse)
}

type ValidateRequest struct{}

type ValidateResponse struct {
	Error *FuncError
}
```

A custom type that was to be used for validating a provider-defined function parameter, _and_ the result from the function could be defined as follows:

```go
func (v CustomStringValue) ValidateParameter(ctx context.Context, req function.ValidateParameterRequest, resp *function.ValidateParameterResponse) {
	if v.IsNull() || v.IsUnknown() {
		return
	}

	if !v.isValid(v.ValueString()) {
		resp.Error = function.NewArgumentFuncError(
			req.Position,
			fmt.Sprintf("Invalid String Value: value %q length does not equal %d", v.ValueString(), 10))
	}
}

func (v CustomStringValue) Validate(ctx context.Context, req function.ValidateRequest, resp *function.ValidateResponse) {
	if v.IsNull() || v.IsUnknown() {
		return
	}

	if !v.isValid(v.ValueString()) {
		resp.Error = function.NewFuncError(
			fmt.Sprintf("Invalid String Value: value %q length does not equal %d", v.ValueString(), 10))
	}
}

func (v CustomStringValue) isValid(in string) bool {
	return len(in) == 10
}
```

The function definition could then be specified as:

```go
func (f ExampleFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
	resp.Definition = function.Definition{
		Parameters: []function.Parameter{
			function.StringParameter{
				CustomType: CustomStringType{},
				Name:       "arg0",
			},
		},
		Return: function.StringReturn{
			CustomType: CustomStringType{},
		},
	}
}
```

The validation of the result value could then be handled by modifying the `ResultData.Set` method:

```go
func (d *ResultData) Set(ctx context.Context, value any) *FuncError {
	reflectValue, reflectDiags := fwreflect.FromValue(ctx, d.value.Type(ctx), value, path.Empty())

	funcErr := FuncErrorFromDiags(ctx, reflectDiags)

	if funcErr != nil {
		return funcErr
	}

	if v, ok := reflectValue.(Validateable); ok {
		resp := ValidateResponse{}

		v.Validate(ctx, ValidateRequest{}, &resp)

		if resp.Error != nil {
			return resp.Error
		}
	}

	d.value = reflectValue

	return nil
}
```

Thoughts?

Reply:
In the case of something like `Validateable`, I'd assume it would end up just returning a plain `error` 🤔. I'm not 100% sure adding a new interface just for validation in reflection is worth it, but I think the general problem has been shown here.

I.e. the `ValidateableAttribute` interface is being used for more than just attributes. Perhaps something that we can't avoid? Something we don't necessarily need to avoid? 🤷🏻 

Would be interested to hear the rest of the teams thoughts on this, although I don't think it's really a showstopper 

Reply:
Thanks for the feedback. Ok, so I've amended the Go doc for `ValidateableAttribute` as follows:

```go
// ValidateableAttribute defines an interface for validating an attribute value.
// The ValidateAttribute method is called implicitly by the framework when value
// types from Terraform are converted into framework types.
type ValidateableAttribute interface {
```

Reply:
LGTM 👍🏻